src/Features/CSharp/Portable/RemoveUnnecessaryLambdaExpression/CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,273 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Utilities;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryLambdaExpression
+{
+    /// <summary>
+    /// DiagnosticAnalyzer that looks code like <c>Goo(() => Bar())</c> and offers to convert it to <c>Goo(Bar)</c>.
+    /// This is only offered on C# 11 and above where this delegate can be cached and will not cause allocations each
+    /// time.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpRemoveUnnecessaryLambdaExpressionDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.RemoveUnnecessaryLambdaExpressionDiagnosticId,
+                   EnforceOnBuildValues.RemoveUnnecessaryLambdaExpression,
+                   CSharpCodeStyleOptions.PreferMethodGroupConversion,
+                   LanguageNames.CSharp,
+                   new LocalizableResourceString(nameof(CSharpFeaturesResources.Remove_unnecessary_lambda_expression), CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources)),
+                   new LocalizableResourceString(nameof(CSharpFeaturesResources.Lambda_expression_can_be_removed), CSharpFeaturesResources.ResourceManager, typeof(CSharpFeaturesResources)),
+                   isUnnecessary: true)
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (((CSharpCompilation)context.Compilation).LanguageVersion >= LanguageVersion.Preview)
```

Comment:
Wasn't there a special value where the compiler team is going to update its usages once C# 11 is released? I can't find it.

cc @jcouv

Reply:
Yes, there's `LanguageVersionFacts.CSharpNext`. But `LanguageVersion.Preview` is okay here too.
When we add `LanguageVersion.CSharp11`, we'll review where `Preview` is used in the produce and we'll fix them up. We just to want hundreds of them (the situation we used to have in tests).

Reply:
You can add a new method to [`LanguageVersionExtensions`](https://github.com/dotnet/roslyn/blob/main/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/LanguageVersionExtensions.cs) to tie this to a particular feature.

Reply:
will do.

Reply:
done.